### ISSUE TITLE
Actually learning the solution?

### ISSUE THREAD

**Issue by pmorerio** [benmoseley/harmonic-oscillator-pinn#1]
Hi, first of all congratulations for the very nice work, the blog-post and the code, I enjoyed them very much!
Not a real issue here, just wanted to share what I found playing around with the code and I think could be interesting.

So basically I restricted the range of x_physiscs , i.e. the range where the physics loss is evaluated, to a smaller interval (e.g. [0,0.6] to see whether the network was able to generalize beyond training data.
Here is what I get:
![image](https://user-images.githubusercontent.com/6783835/142599494-ad322802-eb12-4a65-bc8b-590aa4412f0b.png)
So in my understanding the network is not learning the solution of the differential equation, but only a polynomial approximation which fits the training data only in a specific range. Thus if you would like to use the network to make predictions, you have to make sure that the training data points for the physics loss do cover the range you would like to predict. 

This raises another possible issue, i.e. describing many oscillations of the system may require very high order polynomials and this could go beyond the expressive power of the NN. In the spirit of "physics informed networks" maybe some priors on the solutions could be used to engineer a network which indeed has the capability of parameterize the exact solution, such as periodic activation functions, to mention one.

Best,

P.

**Comment by benmoseley** [comment]
Hi @pmorerio, some really great observations, thanks for trying the code!

Yes, you are exactly right, the PINN will only learn the solution over the range of its "physics loss training points". Basically there is no free lunch: we don't need many training examples of the solution (orange dots), but we do need lots of input points (green dots) sampled over the entire input domain we are interested in solving.

And then the question of whether PINNs are expressive enough to learn many oscillations of this system. It is very likely a small neural network will not be expressive enough, and a larger neural network is needed. Actually, we just wrote a paper on exactly this limitation. In the paper we propose "finite basis physics informed neural networks (FBPINNs)", which combine PINNs with domain decomposition to allow them to scale to large domains, see here: https://arxiv.org/pdf/2107.07871.pdf. In fact, I created an example comparing FBPINNs for PINNs using the harmonic oscillator problem, but for many more oscillations here: https://github.com/benmoseley/FBPINNs/blob/main/examples/Defining%20your%20own%20problem.ipynb.

The idea of using priors on the solution is a very good one. For example, others have used sinusoidal activation functions in the NN as a prior (https://vsitzmann.github.io/siren/). In our FBPINN paper, we show how the boundary conditions of the PDE can automatically be satisfied by using the PINN as part of a mathematical "ansatz" solution too.
